File: class_primer/class_primer__robot_move.cs
using System.Linq;
using static System.Console;

internal class Program
{
    class Robot
    {
        public int Px { get; set; }
        public int Py { get; set; }
        public int Lv { get; set; }

        public Robot( int[] a )
        {
            Px = a[0];
            Py = a[1];
            Lv = a[2];
        }

        public void Move( char d )
        {
            int[] len = { 0, 1, 2, 5, 10 };

            switch( d ) {
            case 'N':
                Py -= len[Lv];
                break;
            case 'S':
                Py += len[Lv];
                break;
            case 'E':
                Px += len[Lv];
                break;
            case 'W':
                Px -= len[Lv];
                break;
            }
            return;
        }
    }

    static void Main()
    {
        int[] HWNK = ReadLine().Split().Select( int.Parse ).ToArray();
        int[] tx = new int[10];
        int[] ty = new int[10];
        Robot[] a = new Robot[HWNK[2]];

        foreach( int i in Enumerable.Range( 0, 10 ) ) {
            int[] t = ReadLine().Split().Select( int.Parse ).ToArray();

            tx[i] = t[0];
            ty[i] = t[1];
        }
        foreach( int i in Enumerable.Range( 0, HWNK[2] ) ) {
            a[i] = new Robot( ReadLine().Split().Select( int.Parse ).ToArray() );
        }
        foreach( int _ in Enumerable.Range( 0, HWNK[3] ) ) {
            string[] s = ReadLine().Split();
            int p = int.Parse( s[0] ) - 1;

            a[p].Move( s[1][0] );
            foreach( int i in Enumerable.Range( 0, tx.Length ) ) {
                if( a[p].Px == tx[i] && a[p].Py == ty[i] ) {
                    a[p].Lv = System.Math.Min( a[p].Lv + 1, 4 );
                    break;
                }
            }
        }
        SetOut( new System.IO.StreamWriter( OpenStandardOutput() ) { AutoFlush = false } );
        foreach( Robot r in a ) {
            WriteLine( $"{r.Px} {r.Py} {r.Lv}" );
        }
        Out.Flush();
        return;
    }
}

File: class_primer/class_primer__robot_move.py
class robot:
    def __init__(self, a):
        self.x = a[0]
        self.y = a[1]
        self.l = a[2]

    def lvlUp(self):
        self.l = min(self.l + 1, 4)

    def move(self, d):
        s = [0, 1, 2, 5, 10]

        if d == "E":
            self.x += s[self.l]
        elif d == "W":
            self.x -= s[self.l]
        elif d == "S":
            self.y += s[self.l]
        elif d == "N":
            self.y -= s[self.l]

    def write(self):
        print(f"{self.x} {self.y} {self.l}")

TL = 10
h, w, n, k = list(map(int, input().split()))
tx = []
ty = []
l = []

for _ in range(TL):
    t = list(map(int, input().split()))

    tx.append(t[0])
    ty.append(t[1])
for _ in range(n):
    l.append(robot(list(map(int, input().split()))))
for _ in range(k):
    t = input().split()
    i = int(t[0]) - 1

    l[i].move(t[1])
    for j in range(TL):
        if l[i].x == tx[j] and l[i].y == ty[j]:
            l[i].lvlUp()
            break
for i in l:
    i.write()
